zebra-chain/src/transaction/sighash.rs

```diff
@@ -0,0 +1,300 @@
+            .personal(&self.personal())
+            .to_state();
+
+        match self.network_upgrade() {
+            Genesis => unreachable!("Zebra checkpoints on Sapling activation"),
+            BeforeOverwinter => unreachable!("Zebra checkpoints on Sapling activation"),
+            Overwinter | Sapling => self
+                .hash_sighash_zip143(&mut hash)
+                .expect("serialization into hasher never fails"),
+            Blossom => unimplemented!(),
+            Heartwood => unimplemented!(),
+            Canopy => unimplemented!(),
+        }
+
+        hash.finalize()
+    }
+
+    fn network_upgrade(&self) -> NetworkUpgrade {
+        NetworkUpgrade::current(self.network, self.height)
+    }
+
+    fn consensus_branch_id(&self) -> ConsensusBranchId {
+        self.network_upgrade()
+            .branch_id()
+            .expect("Zebra checkpoints on Sapling activation")
+    }
+
+    /// Sighash implementation for the overwinter and sapling consensus branches
+    fn hash_sighash_zip143<W: io::Write>(&self, mut writer: W) -> Result<(), io::Error> {
+        self.hash_header(&mut writer)?;
+        self.hash_groupid(&mut writer)?;
+        self.hash_prevouts(&mut writer)?;
+        self.hash_sequence(&mut writer)?;
+        self.hash_outputs(&mut writer)?;
+        self.hash_joinsplits(&mut writer)?;
+        self.hash_lock_time(&mut writer)?;
+        self.hash_expiry_height(&mut writer)?;
+        self.hash_hash_type(&mut writer)?;
+
+        // if let Some((n, script_code, amount)) = transparent_input {
+        //     let mut data = vec![];
+        //     tx.vin[n].prevout.write(&mut data).unwrap();
+        //     script_code.write(&mut data).unwrap();
+        //     data.extend_from_slice(&amount.to_i64_le_bytes());
+        //     (&mut data)
+        //         .write_u32::<LittleEndian>(tx.vin[n].sequence)
+        //         .unwrap();
+        //     h.update(&data);
+        // }
+
+        Ok(())
+    }
+
+    pub(super) fn personal(&self) -> [u8; 16] {
+        let mut personal = [0; 16];
+        (&mut personal[..12]).copy_from_slice(ZCASH_SIGHASH_PERSONALIZATION_PREFIX);
+        (&mut personal[12..])
+            .write_u32::<LittleEndian>(self.consensus_branch_id().into())
+            .unwrap();
+        personal
+    }
+
+    fn hash_header<W: io::Write>(&self, mut writer: W) -> Result<(), io::Error> {
+        writer.write_u32::<LittleEndian>(match &self.trans {
+            Transaction::V1 { .. } => 1,
+            Transaction::V2 { .. } => 2,
+            Transaction::V3 { .. } => 3 | 1 << 31,
+            Transaction::V4 { .. } => 4 | 1 << 31,
+        })
+    }
+
+    fn hash_groupid<W: io::Write>(&self, mut writer: W) -> Result<(), io::Error> {
+        writer.write_u32::<LittleEndian>(match &self.trans {
+            Transaction::V1 { .. } => unimplemented!("value to be determined"),
+            Transaction::V2 { .. } => unimplemented!("value to be determined"),
```

Comment:
V1 and V2 transactions use the bitcoin hashing operations:
https://github.com/zcash/zips/blob/master/zip-0143.rst#motivation
They don't have a group id, so calling this function is a bug, and should panic.

V1 and V2 transactions are rejected after the Overwinter upgrade. So we can apply the transaction version rules at a higher level, and panic here.